### ISSUE TITLE
Unable to build on macOS

### ISSUE THREAD

**Issue by cadenkriese** [ProjectPhysX/OpenCL-Benchmark#1]
Hi! I just tried following the macOS build instructions; I'm getting this output. 

<details>
<summary>./make.sh output</summary>

```
OpenCL-Benchmark on  master [!] in 11s
❯ ./make.sh
In file included from ./src/kernel.cpp:1:
In file included from ./src/kernel.hpp:3:
./src/utilities.hpp:57:86: error: a space is required between consecutive right angle brackets (use '> >')
        double stop() const { return std::chrono::duration_cast<std::chrono::duration<double>>(clock::now()-t).count(); }
                                                                                            ^~
                                                                                            > >
./src/utilities.hpp:570:2: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
        auto words_begin = std::sregex_iterator(s.begin(), s.end(), words_regex);
        ^
./src/utilities.hpp:571:2: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
        auto words_end = std::sregex_iterator();
        ^
./src/kernel.cpp:10:3: error: embedding a #pragma directive within macro arguments is not supported
        #pragma unroll
         ^
./src/kernel.cpp:6:24: note: expansion of macro 'R' requested here
)+"#ifdef cl_khr_fp64"+R( // OpenCL C defines don't work in R() stringification macro
                       ^
./src/kernel.cpp:22:3: error: embedding a #pragma directive within macro arguments is not supported
        #pragma unroll
         ^
./src/kernel.cpp:17:12: note: expansion of macro 'R' requested here
)+"#endif"+R( // cl_khr_fp64
           ^
./src/kernel.cpp:34:3: error: embedding a #pragma directive within macro arguments is not supported
        #pragma unroll
         ^
./src/kernel.cpp:30:24: note: expansion of macro 'R' requested here
)+"#ifdef cl_khr_fp16"+R( // OpenCL C defines don't work in R() stringification macro
                       ^
./src/kernel.cpp:46:3: error: embedding a #pragma directive within macro arguments is not supported
        #pragma unroll
         ^
./src/kernel.cpp:41:12: note: expansion of macro 'R' requested here
)+"#endif"+R( // cl_khr_fp16
           ^
./src/kernel.cpp:57:3: error: embedding a #pragma directive within macro arguments is not supported
        #pragma unroll
         ^
./src/kernel.cpp:41:12: note: expansion of macro 'R' requested here
)+"#endif"+R( // cl_khr_fp16
           ^
./src/kernel.cpp:68:3: error: embedding a #pragma directive within macro arguments is not supported
        #pragma unroll
         ^
./src/kernel.cpp:41:12: note: expansion of macro 'R' requested here
)+"#endif"+R( // cl_khr_fp16
           ^
./src/kernel.cpp:79:3: error: embedding a #pragma directive within macro arguments is not supported
        #pragma unroll
         ^
./src/kernel.cpp:41:12: note: expansion of macro 'R' requested here
)+"#endif"+R( // cl_khr_fp16
           ^
./src/kernel.cpp:91:3: error: embedding a #pragma directive within macro arguments is not supported
        #pragma unroll
         ^
./src/kernel.cpp:41:12: note: expansion of macro 'R' requested here
)+"#endif"+R( // cl_khr_fp16
           ^
./src/kernel.cpp:97:3: error: embedding a #pragma directive within macro arguments is not supported
        #pragma unroll
         ^
./src/kernel.cpp:41:12: note: expansion of macro 'R' requested here
)+"#endif"+R( // cl_khr_fp16
           ^
./src/kernel.cpp:103:3: error: embedding a #pragma directive within macro arguments is not supported
        #pragma unroll
         ^
./src/kernel.cpp:41:12: note: expansion of macro 'R' requested here
)+"#endif"+R( // cl_khr_fp16
           ^
./src/kernel.cpp:109:3: error: embedding a #pragma directive within macro arguments is not supported
        #pragma unroll
         ^
./src/kernel.cpp:41:12: note: expansion of macro 'R' requested here
)+"#endif"+R( // cl_khr_fp16
           ^
2 warnings and 12 errors generated.
In file included from ./src/main.cpp:1:
In file included from ./src/opencl.hpp:11:
./src/utilities.hpp:57:86: error: a space is required between consecutive right angle brackets (use '> >')
        double stop() const { return std::chrono::duration_cast<std::chrono::duration<double>>(clock::now()-t).count(); }
                                                                                            ^~
                                                                                            > >
./src/utilities.hpp:570:2: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
        auto words_begin = std::sregex_iterator(s.begin(), s.end(), words_regex);
        ^
./src/utilities.hpp:571:2: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
        auto words_end = std::sregex_iterator();
        ^
In file included from ./src/main.cpp:1:
./src/opencl.hpp:17:10: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint id = 0u; // unique device ID assigned by get_devices()
                ^
./src/opencl.hpp:20:13: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint memory=0u; // global memory in MB
                   ^
./src/opencl.hpp:21:18: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint memory_used=0u; // track global memory usage in MB
                        ^
./src/opencl.hpp:22:19: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint global_cache=0u, local_cache=0u; // global cache in KB, local cache in KB
                         ^
./src/opencl.hpp:22:35: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint global_cache=0u, local_cache=0u; // global cache in KB, local cache in KB
                                         ^
./src/opencl.hpp:23:24: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint max_global_buffer=0u, max_constant_buffer=0u; // maximum global buffer size in MB, maximum constant buffer size in KB
                              ^
./src/opencl.hpp:23:48: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint max_global_buffer=0u, max_constant_buffer=0u; // maximum global buffer size in MB, maximum constant buffer size in KB
                                                      ^
./src/opencl.hpp:24:20: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint compute_units=0u; // compute units (CUs) can contain multiple cores depending on the microarchitecture
                          ^
./src/opencl.hpp:25:22: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint clock_frequency=0u; // in MHz
                            ^
./src/opencl.hpp:26:13: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        bool is_cpu=false, is_gpu=false;
                   ^
./src/opencl.hpp:26:27: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        bool is_cpu=false, is_gpu=false;
                                 ^
./src/opencl.hpp:27:33: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        bool intel_gpu_above_4gb_patch = false; // memory allocations greater than 4GB need to be specifically enabled on Intel GPUs
                                       ^
./src/opencl.hpp:28:22: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint is_fp64_capable=0u, is_fp32_capable=0u, is_fp16_capable=0u, is_int64_capable=0u, is_int32_capable=0u, is_int16_capable=0u, is_int8_capable=0u;
                            ^
./src/opencl.hpp:28:42: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint is_fp64_capable=0u, is_fp32_capable=0u, is_fp16_capable=0u, is_int64_capable=0u, is_int32_capable=0u, is_int16_capable=0u, is_int8_capable=0u;
                                                ^
./src/opencl.hpp:28:62: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint is_fp64_capable=0u, is_fp32_capable=0u, is_fp16_capable=0u, is_int64_capable=0u, is_int32_capable=0u, is_int16_capable=0u, is_int8_capable=0u;
                                                                    ^
./src/opencl.hpp:28:83: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint is_fp64_capable=0u, is_fp32_capable=0u, is_fp16_capable=0u, is_int64_capable=0u, is_int32_capable=0u, is_int16_capable=0u, is_int8_capable=0u;
                                                                                         ^
./src/opencl.hpp:28:104: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint is_fp64_capable=0u, is_fp32_capable=0u, is_fp16_capable=0u, is_int64_capable=0u, is_int32_capable=0u, is_int16_capable=0u, is_int8_capable=0u;
                                                                                                              ^
./src/opencl.hpp:28:125: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint is_fp64_capable=0u, is_fp32_capable=0u, is_fp16_capable=0u, is_int64_capable=0u, is_int32_capable=0u, is_int16_capable=0u, is_int8_capable=0u;
                                                                                                                                   ^
./src/opencl.hpp:28:145: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint is_fp64_capable=0u, is_fp32_capable=0u, is_fp16_capable=0u, is_int64_capable=0u, is_int32_capable=0u, is_int16_capable=0u, is_int8_capable=0u;
                                                                                                                                                       ^
./src/opencl.hpp:29:12: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint cores=0u; // for CPUs, compute_units is the number of threads (twice the number of cores with hyperthreading)
                  ^
./src/opencl.hpp:30:14: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        float tflops=0.0f; // estimated device FP32 floating point performance in TeraFLOPs/s
                    ^
./src/opencl.hpp:56:69: error: expected expression
                const bool nvidia_192_cores_per_cu = contains_any(to_lower(name), {"gt 6", "gt 7", "gtx 6", "gtx 7", "quadro k", "tesla k"}) || (clock_frequency<1000u&&contains(to_lower(name), "titan")); // identify Kepler GPUs
                                                                                  ^
./src/opencl.hpp:57:68: error: expected expression
                const bool nvidia_64_cores_per_cu = contains_any(to_lower(name), {"p100", "v100", "a100", "a30", " 16", " 20", "titan v", "titan rtx", "quadro t", "tesla t", "quadro rtx"}) && !contains(to_lower(name), "rtx a"); // identify P100, Volta, Turing, A100, A30
                                                                                 ^
./src/opencl.hpp:61:60: error: expected expression
                const float amd = (float)(contains_any(to_lower(vendor), {"amd", "advanced"}))*(is_gpu?(amd_256_cores_per_dualcu?256.0f:amd_128_cores_per_dualcu?128.0f:64.0f):0.5f); // AMD GPUs have 64 cores/CU (GCN, CDNA), 128 cores/dualCU (RDNA, RDNA2) or 256 cores/dualCU (RDNA3), AMD CPUs (with SMT) have 1/2 core/CU
                                                                         ^
./src/opencl.hpp:68:84: error: expected expression
                        if((contains(name, "A770")&&memory>=12602u&&memory<13416u)||(contains_any(name, {"A770", "A750", "A580"})&&memory>=6286u&&memory<6693u)||(contains(name, "A380")&&memory>=4705u&&memory<5010u)) { // 77.5%-82.5% reporting -> /0.8
                                                                                                        ^
./src/opencl.hpp:71:27: error: expected expression
                        if((contains_any(name, {"A770", "0x56a0"})&&memory>=15041u&&memory<15855u)||(contains_any(name, {"A770", "A750", "A580", "0x56a0", "0x56a1", "0x56a2"})&&memory>=7503u&&memory<7910u)||(contains_any(name, {"A380", "0x56a5"})&&memory>=5616u&&memory<5921u)) { // 92.5%-97.5% reporting -> /0.95
                                               ^
./src/opencl.hpp:71:100: error: expected expression
                        if((contains_any(name, {"A770", "0x56a0"})&&memory>=15041u&&memory<15855u)||(contains_any(name, {"A770", "A750", "A580", "0x56a0", "0x56a1", "0x56a2"})&&memory>=7503u&&memory<7910u)||(contains_any(name, {"A380", "0x56a5"})&&memory>=5616u&&memory<5921u)) { // 92.5%-97.5% reporting -> /0.95
                                                                                                                        ^
./src/opencl.hpp:71:207: error: expected expression
                        if((contains_any(name, {"A770", "0x56a0"})&&memory>=15041u&&memory<15855u)||(contains_any(name, {"A770", "A750", "A580", "0x56a0", "0x56a1", "0x56a2"})&&memory>=7503u&&memory<7910u)||(contains_any(name, {"A380", "0x56a5"})&&memory>=5616u&&memory<5921u)) { // 92.5%-97.5% reporting -> /0.95
                                                                                                                                                                                                                                   ^
./src/opencl.hpp:151:14: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        bool exists = false;
                    ^
./src/opencl.hpp:172:23: error: expected expression
                cl_source.push_back({ kernel_code.c_str(), kernel_code.length() });
                                    ^
./src/opencl.hpp:176:32: error: expected expression
                int error = cl_program.build({ info.cl_device }, (build_options+" -w").c_str()); // compile OpenCL C code, disable warnings
                                             ^
./src/opencl.hpp:202:10: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        ulong N = 0ull; // buffer length
                ^
./src/opencl.hpp:203:9: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint d = 1u; // buffer dimensions
               ^
./src/opencl.hpp:204:26: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        bool host_buffer_exists = false;
                                ^
./src/opencl.hpp:205:28: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        bool device_buffer_exists = false;
                                  ^
./src/opencl.hpp:206:28: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        bool external_host_buffer = false;
                                  ^
./src/opencl.hpp:207:17: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T* host_buffer = nullptr; // host buffer
                       ^
./src/opencl.hpp:209:17: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        Device* device = nullptr; // pointer to linked Device
                       ^
./src/opencl.hpp:231:6: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *x=nullptr, *y=nullptr, *z=nullptr, *w=nullptr; // host buffer auxiliary pointers for multi-dimensional array access (array of structures)
            ^
./src/opencl.hpp:231:18: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *x=nullptr, *y=nullptr, *z=nullptr, *w=nullptr; // host buffer auxiliary pointers for multi-dimensional array access (array of structures)
                        ^
./src/opencl.hpp:231:30: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *x=nullptr, *y=nullptr, *z=nullptr, *w=nullptr; // host buffer auxiliary pointers for multi-dimensional array access (array of structures)
                                    ^
./src/opencl.hpp:231:42: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *x=nullptr, *y=nullptr, *z=nullptr, *w=nullptr; // host buffer auxiliary pointers for multi-dimensional array access (array of structures)
                                                ^
./src/opencl.hpp:232:7: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
             ^
./src/opencl.hpp:232:20: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                          ^
./src/opencl.hpp:232:33: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                       ^
./src/opencl.hpp:232:46: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                    ^
./src/opencl.hpp:232:59: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                 ^
./src/opencl.hpp:232:72: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                              ^
./src/opencl.hpp:232:85: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                                           ^
./src/opencl.hpp:232:98: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                                                        ^
./src/opencl.hpp:232:111: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                                                                     ^
./src/opencl.hpp:232:124: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                                                                                  ^
./src/opencl.hpp:232:137: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                                                                                               ^
./src/opencl.hpp:232:150: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                                                                                                            ^
./src/opencl.hpp:232:163: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                                                                                                                         ^
./src/opencl.hpp:232:176: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                                                                                                                                      ^
./src/opencl.hpp:232:189: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                                                                                                                                                   ^
./src/opencl.hpp:232:202: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        T *s0=nullptr, *s1=nullptr, *s2=nullptr, *s3=nullptr, *s4=nullptr, *s5=nullptr, *s6=nullptr, *s7=nullptr, *s8=nullptr, *s9=nullptr, *sA=nullptr, *sB=nullptr, *sC=nullptr, *sD=nullptr, *sE=nullptr, *sF=nullptr;
                                                                                                                                                                                                                ^
./src/opencl.hpp:263:33: warning: rvalue references are a C++11 extension [-Wc++11-extensions]
        inline Memory& operator=(Memory&& memory) noexcept { // move assignment
                                       ^
./src/opencl.hpp:263:43: error: expected ';' at end of declaration list
        inline Memory& operator=(Memory&& memory) noexcept { // move assignment
                                                 ^
                                                 ;
./src/opencl.hpp:221:38: error: use of undeclared identifier 'capacity'
                        device.info.memory_used += (uint)(capacity()/1048576ull); // track device memory usage
                                                          ^
./src/opencl.hpp:222:163: error: use of undeclared identifier 'capacity'
                        if(device.info.memory_used>device.info.memory) print_error("Device \""+device.info.name+"\" does not have enough memory. Allocating another "+to_string((uint)(capacity()/1048576ull))+" MB would use a total of "+to_string(device.info.memory_used)+" MB / "+to_string(device.info.memory)+" MB.");
                                                                                                                                                                                       ^
./src/opencl.hpp:224:124: error: use of undeclared identifier 'capacity'
                        device_buffer = cl::Buffer(device.get_cl_context(), CL_MEM_READ_WRITE|((int)device.info.intel_gpu_above_4gb_patch<<23), capacity(), nullptr, &error); // for Intel GPUs, set flag CL_MEM_ALLOW_UNRESTRICTED_SIZE_INTEL = (1<<23)
                                                                                                                                                ^
./src/opencl.hpp:225:79: error: use of undeclared identifier 'capacity'
                        if(error==-61) print_error("Memory size is too large at "+to_string((uint)(capacity()/1048576ull))+" MB. Device \""+device.info.name+"\" accepts a maximum buffer size of "+to_string(device.info.max_global_buffer)+" MB.");
                                                                                                   ^
./src/opencl.hpp:245:3: error: use of undeclared identifier 'write_to_device'
                write_to_device();
                ^
./src/opencl.hpp:257:3: error: use of undeclared identifier 'write_to_device'
                write_to_device();
                ^
./src/opencl.hpp:261:3: error: use of undeclared identifier 'delete_buffers'
                delete_buffers();
                ^
./src/opencl.hpp:445:10: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        ulong N = 0ull; // kernel range
                ^
./src/opencl.hpp:446:28: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
        uint number_of_parameters = 0u;
                                  ^
./src/opencl.hpp:459:25: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
        template<class T, class... U> inline void link_parameters(const uint starting_position, const T& parameter, const U&... parameters) {
                               ^
./src/opencl.hpp:464:16: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
        template<class... T> inline Kernel(const Device& device, const ulong N, const string& name, const T&... parameters) { // accepts Memory<T> objects and fundamental data type constants
                      ^
./src/opencl.hpp:471:16: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
        template<class... T> inline Kernel(const Device& device, const ulong N, const uint workgroup_size, const string& name, const T&... parameters) { // accepts Memory<T> objects and fundamental data type constants
                      ^
./src/opencl.hpp:487:16: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
        template<class... T> inline Kernel& add_parameters(const T&... parameters) { // add parameters to the list of existing parameters
                      ^
./src/opencl.hpp:491:16: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
        template<class... T> inline Kernel& set_parameters(const uint starting_position, const T&... parameters) { // set parameters starting at specified position
                      ^
./src/main.cpp:8:14: warning: range-based for loop is a C++11 extension [-Wc++11-extensions]
        for(float v : values) {
                    ^
./src/main.cpp:154:10: error: no member named 'write_to_device' in 'Memory<float>'
                buffer.write_to_device();
                ~~~~~~ ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
60 warnings and 20 errors generated.
./make.sh: line 10: ./bin/OpenCL-Benchmark: No such file or directory
```
</details>

I'm not really familiar with C++ build systems, so I apologize if this is a trivial mistake. I just wanted to use this project to test performance differences between Apple's OpenCL driver and that of Asahi Linux.

**Comment by ProjectPhysX** [comment]
Hi @cadenkriese,

apologies, there was a few issues with macOS. Thanks for the build error messages! macOS seems to need C++17, and `pragma unroll` is not supported. I just patched these issues in v1.1. Please let me know if there is further problems!

Regards,
Moritz

**Comment by cadenkriese** [comment]
Thanks! All good now. Here are my results in case you, or anyone else, is interested. Asahi doesn't support OpenCL 1.2 (yet), but I'll test it there once they do.

```
.-----------------------------------------------------------------------------.
|----------------.------------------------------------------------------------|
| Device ID    0 | Apple M1 Max                                               |
|----------------'------------------------------------------------------------|
|----------------.------------------------------------------------------------|
| Device ID      | 0                                                          |
| Device Name    | Apple M1 Max                                               |
| Device Vendor  | Apple                                                      |
| Device Driver  | 1.2 1.0                                                    |
| OpenCL Version | OpenCL C 1.2                                               |
| Compute Units  | 32 at 1000 MHz (4096 cores, 8.192 TFLOPs/s)                |
| Memory, Cache  | 21845 MB, 0 KB global / 32 KB local                        |
| Buffer Limits  | 4096 MB global, 1048576 KB constant                        |
|----------------'------------------------------------------------------------|
| Info: OpenCL C code successfully compiled.                                  |
UNSUPPORTED (log once): createKernel: newComputePipelineState failed
| FP64  compute                                          not supported        |
| FP32  compute                                         2.484 TFLOPs/s (1/3 ) |
| FP16  compute                                          not supported        |
| INT64 compute                                         1.338  TIOPs/s (1/8 ) |
| INT32 compute                                         2.425  TIOPs/s (1/3 ) |
| INT16 compute                                         2.550  TIOPs/s (1/3 ) |
| INT8  compute                                         2.528  TIOPs/s (1/3 ) |
| Memory Bandwidth ( coalesced read      )                        229.10 GB/s |
| Memory Bandwidth ( coalesced      write)                        294.03 GB/s |
| Memory Bandwidth (misaligned read      )                        355.78 GB/s |
| Memory Bandwidth (misaligned      write)                        376.65 GB/s |
| PCIe   Bandwidth (send                 )                         88.60 GB/s |
| PCIe   Bandwidth (   receive           )                         88.17 GB/s |
| PCIe   Bandwidth (        bidirectional)            (Gen4 x16)   76.66 GB/s |
|-----------------------------------------------------------------------------|
|-----------------------------------------------------------------------------|
| Done. Press Enter to exit.                                                  |
'-----------------------------------------------------------------------------'```
